dsl_parser/tests/test_constraints.py

```diff
@@ -0,0 +1,311 @@
+
+
+class TestLessOrEqual(testtools.TestCase):
+
+    def test_successful(self):
+        dummy = cnstrs.LessOrEqual(3)
+        self.assertTrue(dummy.predicate(2))
+        self.assertFalse(dummy.predicate(4))
+
+    def test_raises_error(self):
+        class A(object):
+            def __le__(self, other):
+                raise TypeError()
+
+        dummy = cnstrs.LessOrEqual(3)
+        self.assertRaises(exceptions.ConstraintException, dummy.predicate, A())
+
+
+class TestInRange(testtools.TestCase):
+
+    def test_successful(self):
+        dummy = cnstrs.InRange([0, 5])
+        self.assertTrue(dummy.predicate(2))
+        self.assertFalse(dummy.predicate(-4))
+
+    def test_raises_error(self):
+        class A(object):
+            def __ge__(self, item):
+                raise TypeError()
+
+            def __le__(self, item):
+                raise TypeError()
+
+        dummy = cnstrs.InRange([0, 5])
+        self.assertRaises(exceptions.ConstraintException, dummy.predicate, A())
+
+
+class TestValidValues(testtools.TestCase):
+
+    def test_successful(self):
+        dummy = cnstrs.ValidValues([0, 2])
+        self.assertTrue(dummy.predicate(0))
+        self.assertFalse(dummy.predicate(1))
+
+    def test_raises_error(self):
+        class A(object):
+            def __eq__(self, item):
+                raise TypeError()
+
+        dummy = cnstrs.ValidValues([0, 2])
+        self.assertRaises(exceptions.ConstraintException, dummy.predicate, A())
+
+
+class TestLength(testtools.TestCase):
+
+    def test_successful(self):
+        dummy = cnstrs.Length(1)
+        self.assertTrue(dummy.predicate([0]))
+        self.assertFalse(dummy.predicate([]))
+
+    def test_raises_error(self):
+        class A(object):
+            def __len__(self):
+                raise TypeError()
+
+        dummy = cnstrs.Length(2)
+        self.assertRaises(exceptions.ConstraintException, dummy.predicate, A())
+
+
+class TestMinLength(testtools.TestCase):
+
+    def test_successful(self):
+        dummy = cnstrs.MinLength(1)
+        self.assertTrue(dummy.predicate([0, 1]))
+        self.assertFalse(dummy.predicate([]))
+
+    def test_raises_error(self):
+        class A(object):
+            def __len__(self):
+                raise TypeError()
+
+        dummy = cnstrs.MinLength(2)
+        self.assertRaises(exceptions.ConstraintException, dummy.predicate, A())
+
+
+class TestMaxLength(testtools.TestCase):
+
+    def test_successful(self):
+        dummy = cnstrs.MaxLength(1)
+        self.assertTrue(dummy.predicate([0]))
+        self.assertFalse(dummy.predicate([0, 1]))
+
+    def test_raises_error(self):
+        class A(object):
+            def __len__(self):
+                raise TypeError()
+
+        dummy = cnstrs.MaxLength(1)
+        self.assertRaises(exceptions.ConstraintException, dummy.predicate, A())
+
+
+class TestPattern(testtools.TestCase):
+
+    def test_successful(self):
```

Comment:
maybe test more complex patterns?
Also what about lang that are no English

Reply:
Do we need to add some specialized support if `re.match(..)` doesn't support that?
The Pattern.predicate function is just a `re.match(..)` call.